src/types.h

```diff
@@ -353,7 +353,7 @@ inline Square make_square(File f, Rank r) {
 }
 
 inline Piece make_piece(Color c, PieceType pt) {
-  return Piece((c << 3) | pt);
+  return Piece((c << 3) + pt);
```

Comment:
I already explained why I changed this. It allows the compiler to perform the addition implicitly by an indexed addressing mode instead of being forced to perform an or.

Basically, any code that now implements [us][pt] will also implement [make_piece(us, pt)], because [us][pt] really means [(us << 3) + pt]. So, if the compiler is any good, [make_piece(us, pt)] will never be less efficient than the current code. But if make_piece() uses an OR, the compiler might be forced to generate something less efficient.

I did not look at the assembly output of gcc, but creating more opportunities for optimisation seems like a good thing where it does not decrease readability. And a + is not less readable than an OR, I would say.

But OK, let's have a look at the assembly output. When performing a castling move, do_move() now needs to create a rook of the right color using make_piece(). With an OR in make_piece(), the assembly code (AT&T syntax) contains:
        movl 40(%rsp), %eax  ; %eax = us
        leal 0(,%rax,8), %r11d ; %r11d = 8 * us
        movl %r11d, %eax      ; %eax = 8 * us
        orl $4, %eax             ; %eax = 8 * us | ROOK = make_piece(us, ROOK)
        cmpl %r13d, 48(%rsp) ; compare to and from squares (do_castling())
with a +, the corresponding generated code reads:
        movl 20(%rsp), %eax  ; %eax = us
        cmpl %r13d, 32(%rsp)  ; compare to and from squares (do_castling())
        leal 0(,%rax,8), %r11d  ; %r11d = 8 * us
        leal 4(%r11), %eax      ; %eax = 8 * us + ROOK = make_piece(us, ROOK)
So we already save 1 instruction here. Note that the compiler could also have simply replaced the orl with an addl, if there had not been a more efficient way of translating the addition.

I think the second piece of code could be one instruction shorter by doing leal 4(,%rax,8),%eax, but the intermediate value in %r11 is needed again later. This would not have been possible in the first piece of code, because you can't do an OR by adding a displacement to a leal instruction.

That + in make_piece() leads to more efficient code than OR is confirmed by the fact that do_move allocates 104 bytes on the stack when using + and 136 bytes when using OR. Apparently it reduces register pressure.

So going to back to OR will hurt speed, but some gain from the patch will probably remain. Just let me know.